src/libmongoc/tests/test-mongoc-server-stream.c

```diff
@@ -0,0 +1,227 @@
+
+   /* Re-use client_opmsg, to ensure it still uses opmsg. */
+   future = future_client_command_simple (client_opmsg,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   request = mock_server_receives_msg (server, 0, tmp_bson ("{'ping': 1}"));
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_MSG);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Check that selecting the server still returns the OP_MSG server */
+   sd = mongoc_client_select_server (
+      client_opquery, true /* for writes */, NULL /* read prefs */, &error);
+   ASSERT_OR_PRINT (sd, error);
+   ASSERT_MATCH (mongoc_server_description_hello_response (sd),
+                 "{'maxWireVersion': 6}");
+   mongoc_server_description_destroy (sd);
+
+   mock_server_destroy (server);
+   mongoc_uri_destroy (uri);
+   mongoc_client_pool_push (pool, client_opquery);
+   mongoc_client_pool_push (pool, client_opmsg);
+   mongoc_client_pool_destroy (pool);
+}
+
+/* Test that a connection uses the server description from the handshake when
+ * checking wire version (instead of the server description from the topology
+ * description). */
+static void
+test_server_stream_ties_server_description_single (void *unused)
+{
+   mongoc_client_t *client;
+   mongoc_uri_t *uri;
+   mock_server_t *server;
+   request_t *request;
+   future_t *future;
+   bson_error_t error;
+   mongoc_server_description_t *sd;
+
+   server = mock_server_new ();
+   mock_server_run (server);
+   uri = mongoc_uri_copy (mock_server_get_uri (server));
+   client = mongoc_client_new_from_uri (uri);
+
+   /* Create a connection on client. */
+   future = future_client_command_simple (client,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   /* The first command on a client creates a new connection. */
+   request = mock_server_receives_legacy_hello (server, NULL);
+   mock_server_replies_simple (request, HELLO_POST_OPMSG);
+   request_destroy (request);
+   /* Check that the mock server receives an OP_MSG. */
+   request = mock_server_receives_msg (server, 0, tmp_bson ("{'ping': 1}"));
+   ASSERT_CMPINT ((int) request->opcode, ==, (int) MONGOC_OPCODE_MSG);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* Muck with the topology description. */
+   /* Pass in an empty error. */
+   memset (&error, 0, sizeof (bson_error_t));
```

Comment:
Is there any reason `error` wouldn't still be empty? Is this just being defensive in case something _did_ set it earlier and you happened not to fail via an assertion? Just to confirm, this is the _only_ function in this test where `error` is an _input_ parameter, correct?

Alternatively, I suppose you could declare `bson_error_t empty_error = {0}` for this function to only be used for this purpose.

Reply:
> Is this just being defensive in case something did set it earlier and you happened not to fail via an assertion?

No, this is so an uninitialized error is not passed in. I updated the comment to note that the error should be zeroed out.

> Just to confirm, this is the only function in this test where error is an input parameter, correct?

Correct.